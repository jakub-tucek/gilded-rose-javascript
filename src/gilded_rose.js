export class Item {
  constructor(name, sellIn, quality){
    this.name = name;
    this.sellIn = sellIn;
    this.quality = quality;
  }
}

const rules = {
  'Aged Brie': {
    updateQuality(item) {
      if (item.quality < 50) {
        if (item.sellIn <= 0) {
          item.quality += 2;
        } else {
          item.quality += 1;
        }
      }
    },
  },

  'Sulfuras, Hand of Ragnaros': {
    updateQuality() {},

    updateSellIn() {},
  },

  'Backstage passes to a TAFKAL80ETC concert': {
    updateQuality(item) {
      if (item.sellIn > 0) {
        if (item.sellIn <= 5) {
          item.quality += 3;
        } else if (item.sellIn <= 10) {
          item.quality += 2;
        } else {
          item.quality += 1;
        }
      } else {
        debugger;

        item.quality = 0;
      }

      if (item.quality > 50) {
        item.quality = 50;
      }
    },
  },

  default: {
    updateQuality(item) {
      if (item.quality) {
        if (item.sellIn > 0) {
          item.quality -= 1;
        } else if (item.sellIn <= 0) {
          item.quality -= 2;
        }
      }
    },

    updateSellIn(item) {
      item.sellIn -= 1;
    },
  },
};

export class Shop {
  constructor(items=[]){
    this.items = items;
  }

  updateQuality() {
    this.items.forEach((item) => {
      const rule = rules[item.name] || {};

      const updateQuality = rule.updateQuality || rules.default.updateQuality;
      const updateSellIn = rule.updateSellIn || rules.default.updateSellIn;

      updateQuality(item);
      updateSellIn(item);
    });

    return this.items;
  }
}
