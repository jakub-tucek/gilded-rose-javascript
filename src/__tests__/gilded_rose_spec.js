import { Shop, Item } from '../gilded_rose';

describe("Gilded Rose", function () {
  let items;

  beforeEach(() => {
    items = [
      new Item('+5 Dexterity Vest', 10, 20),
      new Item('Aged Brie', 2, 0),
      new Item('Aged Brie', 2, 50),
      new Item('Elixir of the Mongoose', 5, 7),
      new Item('Sulfuras, Hand of Ragnaros', 0, 80),
      new Item('Sulfuras, Hand of Ragnaros', -1, 80),
      new Item('Backstage passes to a TAFKAL80ETC concert', 15, 20),
      new Item('Backstage passes to a TAFKAL80ETC concert', 10, 49),
      new Item('Backstage passes to a TAFKAL80ETC concert', 5, 49),
      new Item('Backstage passes to a TAFKAL80ETC concert', 5, 40),
    ];
  });

  it('has correct state after each of 6 days', () => {
    const gildedRose = new Shop(items);

    for (let i = 0; i < 7; i++) {
      gildedRose.updateQuality();

      expect(gildedRose.items).toMatchSnapshot();
    }
  });
});
