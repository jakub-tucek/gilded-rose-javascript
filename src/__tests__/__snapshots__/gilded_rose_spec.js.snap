// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gilded Rose has correct state after each of 6 days 1`] = `
Array [
  Item {
    "name": "+5 Dexterity Vest",
    "quality": 19,
    "sellIn": 9,
  },
  Item {
    "name": "Aged Brie",
    "quality": 1,
    "sellIn": 1,
  },
  Item {
    "name": "Aged Brie",
    "quality": 50,
    "sellIn": 1,
  },
  Item {
    "name": "Elixir of the Mongoose",
    "quality": 6,
    "sellIn": 4,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": 0,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": -1,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 21,
    "sellIn": 14,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 9,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 4,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 43,
    "sellIn": 4,
  },
]
`;

exports[`Gilded Rose has correct state after each of 6 days 2`] = `
Array [
  Item {
    "name": "+5 Dexterity Vest",
    "quality": 18,
    "sellIn": 8,
  },
  Item {
    "name": "Aged Brie",
    "quality": 2,
    "sellIn": 0,
  },
  Item {
    "name": "Aged Brie",
    "quality": 50,
    "sellIn": 0,
  },
  Item {
    "name": "Elixir of the Mongoose",
    "quality": 5,
    "sellIn": 3,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": 0,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": -1,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 22,
    "sellIn": 13,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 8,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 3,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 46,
    "sellIn": 3,
  },
]
`;

exports[`Gilded Rose has correct state after each of 6 days 3`] = `
Array [
  Item {
    "name": "+5 Dexterity Vest",
    "quality": 17,
    "sellIn": 7,
  },
  Item {
    "name": "Aged Brie",
    "quality": 4,
    "sellIn": -1,
  },
  Item {
    "name": "Aged Brie",
    "quality": 50,
    "sellIn": -1,
  },
  Item {
    "name": "Elixir of the Mongoose",
    "quality": 4,
    "sellIn": 2,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": 0,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": -1,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 23,
    "sellIn": 12,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 7,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 2,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 49,
    "sellIn": 2,
  },
]
`;

exports[`Gilded Rose has correct state after each of 6 days 4`] = `
Array [
  Item {
    "name": "+5 Dexterity Vest",
    "quality": 16,
    "sellIn": 6,
  },
  Item {
    "name": "Aged Brie",
    "quality": 6,
    "sellIn": -2,
  },
  Item {
    "name": "Aged Brie",
    "quality": 50,
    "sellIn": -2,
  },
  Item {
    "name": "Elixir of the Mongoose",
    "quality": 3,
    "sellIn": 1,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": 0,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": -1,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 24,
    "sellIn": 11,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 6,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 1,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 1,
  },
]
`;

exports[`Gilded Rose has correct state after each of 6 days 5`] = `
Array [
  Item {
    "name": "+5 Dexterity Vest",
    "quality": 15,
    "sellIn": 5,
  },
  Item {
    "name": "Aged Brie",
    "quality": 8,
    "sellIn": -3,
  },
  Item {
    "name": "Aged Brie",
    "quality": 50,
    "sellIn": -3,
  },
  Item {
    "name": "Elixir of the Mongoose",
    "quality": 2,
    "sellIn": 0,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": 0,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": -1,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 25,
    "sellIn": 10,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 5,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 0,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 0,
  },
]
`;

exports[`Gilded Rose has correct state after each of 6 days 6`] = `
Array [
  Item {
    "name": "+5 Dexterity Vest",
    "quality": 14,
    "sellIn": 4,
  },
  Item {
    "name": "Aged Brie",
    "quality": 10,
    "sellIn": -4,
  },
  Item {
    "name": "Aged Brie",
    "quality": 50,
    "sellIn": -4,
  },
  Item {
    "name": "Elixir of the Mongoose",
    "quality": 0,
    "sellIn": -1,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": 0,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": -1,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 27,
    "sellIn": 9,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 4,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 0,
    "sellIn": -1,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 0,
    "sellIn": -1,
  },
]
`;

exports[`Gilded Rose has correct state after each of 6 days 7`] = `
Array [
  Item {
    "name": "+5 Dexterity Vest",
    "quality": 13,
    "sellIn": 3,
  },
  Item {
    "name": "Aged Brie",
    "quality": 12,
    "sellIn": -5,
  },
  Item {
    "name": "Aged Brie",
    "quality": 50,
    "sellIn": -5,
  },
  Item {
    "name": "Elixir of the Mongoose",
    "quality": 0,
    "sellIn": -2,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": 0,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": -1,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 29,
    "sellIn": 8,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 3,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 0,
    "sellIn": -2,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 0,
    "sellIn": -2,
  },
]
`;
